### ISSUE TITLE
Inconsistent quantile query results

### ISSUE THREAD

**Issue by gmwiku** [VictoriaMetrics/VictoriaMetrics#1612]
**Describe the bug**
The same data, the same query but different results

**To Reproduce**
```
#! /bin/bash

docker run -d --rm --name victoriametrics -p "8428:8428" victoriametrics/victoria-metrics -search.maxPointsPerTimeseries=3000000

start=1630454400
step=60
repetitions=10000
nanosuffix="000000000"

sleep 3

number=0
timestamp=$start
until [ $number -gt $repetitions ] ; do
  curl -s -d "measurement,tag1=value1,tag2=value2 field=$number $timestamp$nanosuffix" -X POST 'http://localhost:8428/write'
  number=$((number+1))
  timestamp=$((timestamp+$step))
done
end=$(($timestamp-$step))

sleep 3

echo "start: $start   end: $end   step: $step   range: $(($end-$start))"

query="query=quantile_over_time(0.95,avg_over_time(measurement_field{}[300s])[$(($end-$start))s])"
echo "$query"
curl -s -G http://localhost:8428/api/v1/query_range -d "$query&start=$start&end=$end&step=$step" | jq '.' | tail -8 | head -2
curl -s -G http://localhost:8428/api/v1/query_range -d "$query&start=$start&end=$end&step=$step" | jq '.' | tail -8 | head -2

docker rm -f victoriametrics

```

**Expected behavior**
I've expected the same result. If the query is changed to
`query="query=avg_over_time(avg_over_time(measurement_field{}[300s])[$(($end-$start))s])"`
result is repetitable.
In case we change query to
`query="query=quantile_over_time(0.95,measurement_field{}[$(($end-$start))s])"`
result is also repetitable, but I'm not sure if valid - I would expect result (whole, not just the last item) will be increasing as source data are.

**Logs**
N/A

**Screenshots**
N/A

**Version**
Latest docker image
```
victoriametrics/victoria-metrics   latest    5019c08657b0   11 days ago   24.4MB
```

**Used command-line flags**
see at reproduce script

**Comment by valyala** [comment]
VictoriaMetrics uses [randomized algorithm](https://en.m.wikipedia.org/wiki/Randomized_algorithm) when calculating `quantile_over_time` over `N` samples where `N>1000`. See https://github.com/valyala/histogram/blob/a43f280016145ea9cda8253c881284461eb86cbb/histogram.go#L55 . This may result in slightly different responses for repeated queries. The randomized algorithm allowes limiting memory usage when calculating quantiles over big number of samples. The algorithm can converted to deterministic by initializing the pseudo-random number generator inside `histogram.Fast` datastructure with the hardcoded state. I'll follow this path in the upcoming commit.

P.S. Thanks for the reprode script! It is awesome!

**Comment by valyala** [comment]
oops, the issue shouldn't be closed until it is fixed in the next release.

The issue should be fixed in the commit 68ea3d18f7645f08212d21d3555730ea7f91ca47 . @gmwiku , could you build VictoriaMetrics from this commit according to [these instructions](https://docs.victoriametrics.com/#how-to-build-from-sources) and verify whether this fixes the issue?

**Comment by gmwiku** [comment]
thank you @valyala for very quick response! commit [68ea3d1] is very close to fixing the issue.

reproduce script gives the result 9533 - now it's repeatable and slightly (~3%) above expected 9500 value. 

I've build also another VM docker image with maxSamples value increased to 10000 and final result 9498 is much more satisfying for me, even if it uses more resources for calculations. Is it possible to add option to parametrize this value?? It would be great!

**Comment by valyala** [comment]
> reproduce script gives the result 9533 - now it's repeatable and slightly (~3%) above expected 9500 value.

The source of the error is the algorithm that selects a random set of samples and then calculates the quantile on top of the set. Obviously, the error changes with each random selection. It works OK in most practical cases, but it has no upper bound in pathological cases.

> I've build also another VM docker image with maxSamples value increased to 10000 and final result 9498 is much more satisfying for me, even if it uses more resources for calculations. Is it possible to add option to parametrize this value?? It would be great!

I think it would be better adding `exact_quantile` function for calculating exact quantiles instead of estimated quantiles. Another option is to add `estimated_quantile(phi, epsilon, q)` function, where the estimation error bound could be controlled with the `epsilon` arg.

**Comment by valyala** [comment]
FYI, [VictoriaMetrics v1.66.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases) should prevent from non-constant results when `quantile_over_time()` function is called multiple times on the same time range for the same set of time series.

**Comment by valyala** [comment]
> Do you know how it is implemented in Prometheus? if it is uses randomized algorithm I would prefer adding exact_quantile, otherwise I would prefer implementing estimated_quantile and change quantile to use exact algorithm.

Prometheus calculates exact quantile with some interpolation if the quantile points to the middle of two values. For example, median over two values returns mean value instead. See [this issue](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/1625) for details. We decided implementing the same algorithm at https://github.com/VictoriaMetrics/VictoriaMetrics/pull/1646 . It will be included in the next release of VictoriaMetrics.

**Comment by valyala** [comment]
@gmwiku , could build VictoriaMetrics from the commit c4c77aa2dd1eded64b0506a00994aeb8b600a647 according to [these docs](https://docs.victoriametrics.com/#how-to-build-from-sources) and verify whether it calculates `quantile_over_time()` in the expected way? This commit will be included in the next release of VictoriaMetrics.

**Comment by gmwiku** [comment]
> @gmwiku , could [...] verify whether it calculates `quantile_over_time()` in the expected way? [...]
Perfect for me. :)

**Comment by valyala** [comment]
FYI, the bugfix has been included in the release [v1.67.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).
